src/sink_http.c

```diff
@@ -156,76 +178,246 @@ static int add_metrics(void* data,
+        pre_len = strlen(prefix);
+    }
+
+    /*
+     * Find the json object mapped to the partition string.
+     * Create one if there isn't one yet.
+     */
+    json_t* obj = NULL;
+    if (hashmap_get(json_map, partition, (void**)&obj) == -1) {
+        /* Create a new json object if it doesn't exist */
+        obj = json_object();
+        hashmap_put(json_map, partition, obj);
+    }
+
+    assert(obj != NULL);
+    _add_metric_to_json(obj, type, prefix, pre_len, metric, name_len, value, config);
     return 0;
 }
 
-static int json_cb(const char* buf, size_t size, void* d) {
+/*
+ * Callback function for dumping json object to string.
+ * @arg buf The data buffer to dump.
+ * @arg size Number of bytes to dump.
+ * @arg buf The output buffer.
+ * @return 0 on success.
+ */
+static int _json_cb(const char* buf, size_t size, void* d) {
     strbuf* sbuf = (strbuf*)d;
     strbuf_cat(sbuf, buf, size);
     return 0;
 }
 
-static int serialize_metrics(struct http_sink* sink, metrics* m, void* data) {
-    json_t* jobject = json_object();
-    const sink_config_http* httpconfig = (const sink_config_http*)sink->sink.sink_config;
+/*
+ * Escape a string and add it to the post buffer.
+ * @arg post_buf The post buffer.
+ * @arg value The string to escape and append.
+ * @arg len The length of the string to append. 0 if the full string should be appended.
+ * @arg curl The CURL object for string escaping.
+ */
+static void _escape_and_append_to_buffer(strbuf* post_buf,
+                                         const char* value,
+                                         int len,
+                                         CURL* curl) {
+    char* escaped_data = curl_easy_escape(curl, value, len);
+    strbuf_cat(post_buf, escaped_data, strlen(escaped_data));
+    curl_free(escaped_data);
+}
 
-    struct cb_info info = {
-        .jobject = jobject,
-        .config = sink->sink.global_config,
-        .httpconfig = httpconfig
-    };
-    /* produce a metrics json object */
-    metrics_iter(m, &info, add_metrics);
+/*
+ * Verify if a key exists in a query string or not. This function simply scans
+ * through the query string, hence it shouldn't be called often.
+ * @arg key The key to search. It should not contain characters '&' and '='.
+ * @arg partition_string The partition string to search in.
+ * @return True if the key exists as a key name(as key in key=123 but not in key123=123)
+ *         in the query string, false otherwise.
+ */
+bool _key_exists_in_query_string(const char* key,
+                                 const char* query_string) {
+    if (key == NULL || query_string == NULL) {
+        return false;
+    }
+
+    int len = strlen(key);
+    do {
+        if (strncmp(query_string, key, len) == 0 &&
+            (query_string[len] == '=' || query_string[len] == '&' || query_string[len] == '\0')) {
+            return true;
+        }
+        query_string = strchr(query_string + 1, '&');
```

Comment:
Is it possible to have ```query_string``` as an empty string (but with a larger allocation size)? If so, this would run over.